File: lib/target.ml
type t = string

let to_string target = target

let of_string target =
  match String.split_on_char '#' target with
  | [ flake_ref; config_name ] when flake_ref = "" || config_name = "" ->
      Error
        (`Msg
           "both flake reference and config name are required in --target \
            <flake-ref>#<config-name>")
  | [ _flake_ref; _config_name ] -> Ok target
  | _ -> Error (`Msg "--target must use <flake-ref>#<config-name>")

type hooks_config = {
  post_launch : string list;
  pre_stop : string list;
}

type descriptor = {
  kernel : string;
  disk : string;
  initrd : string option;
  cmdline : string;
  cpus : int;
  memory_mib : int;
  configured_users : string list;
  hooks : hooks_config;
}

let default_cmdline = "console=ttyS0 root=/dev/vda2 ro"


let descriptor_of_json json =
  let open Yojson.Basic.Util in
  let str key default = json |> member key |> to_string_option |> Option.value ~default in
  let int key default = json |> member key |> to_int_option |> Option.value ~default in
  let kernel = str "kernel" "" in
  let disk = str "disk" "" in
  let initrd = json |> member "initrd" |> to_string_option in
  let cmdline = str "cmdline" default_cmdline in
  let cpus = int "cpus" 1 in
  let memory_mib = int "memory_mib" 1024 in
  let configured_users =
    match json |> member "configuredUsers" with
    | `List items -> List.filter_map to_string_option items
    | _ -> []
  in
  let parse_hook_set hooks_json key =
    match hooks_json |> member key with
    | `Assoc pairs ->
        pairs
        |> List.sort (fun (k1, _) (k2, _) -> String.compare k1 k2)
        |> List.filter_map (fun (_, v) -> to_string_option v)
    | `List items -> List.filter_map to_string_option items
    | _ -> []
  in
  let hooks =
    match json |> member "hooks" with
    | `Assoc _ as hooks_json ->
        { post_launch = parse_hook_set hooks_json "post-launch";
          pre_stop = parse_hook_set hooks_json "pre-stop" }
    | _ -> { post_launch = []; pre_stop = [] }
  in
  { kernel; disk; initrd; cmdline; cpus; memory_mib; configured_users; hooks }

type resolution_error = {
  target : string;
  details : string;
  exit_code : int option;
}

let expand_flake_ref_tilde target =
  match String.index_opt target '#' with
  | None -> Config.expand_tilde target
  | Some i ->
      let flake_ref = String.sub target 0 i in
      let rest = String.sub target i (String.length target - i) in
      Config.expand_tilde flake_ref ^ rest

let store_root_of_path path =
  let prefix = "/nix/store/" in
  if not (String.starts_with ~prefix path) then None
  else
    let rest =
      String.sub path (String.length prefix)
        (String.length path - String.length prefix)
    in
    match String.index_opt rest '/' with
    | None -> Some path
    | Some index -> Some (prefix ^ String.sub rest 0 index)

let ensure_store_realized path =
  match store_root_of_path path with
  | None -> ()
  | Some store_root ->
      if Sys.file_exists store_root then ()
      else
        let _ =
          Process.run ~prog:"nix-store" ~args:[ "--realise"; store_root ] ()
        in
        ()

let split_target target =
  match String.index_opt target '#' with
  | None -> None
  | Some hash_index ->
      let flake_ref = String.sub target 0 hash_index in
      let config_name =
        String.sub target (hash_index + 1)
          (String.length target - hash_index - 1)
      in
      if flake_ref = "" || config_name = "" then None
      else Some (flake_ref, config_name)

let canonicalize_target target =
  match split_target target with
  | None -> target
  | Some (flake_ref, config_name) ->
      if String.starts_with ~prefix:"nixosConfigurations." config_name then
        target
      else flake_ref ^ "#nixosConfigurations." ^ config_name

let check_target_exists target =
  let result =
    Process.run ~prog:"nix"
      ~args:[ "eval"; target; "--apply"; "x: true" ]
      ()
  in
  if result.status <> 0 then
    Error
      {
        target;
        details = Printf.sprintf "target '%s' not found in flake" target;
        exit_code = Some result.status;
      }
  else Ok ()

let resolve_descriptor target =
  let target = expand_flake_ref_tilde target in
  let target = canonicalize_target target in
  match Sys.getenv_opt "EPI_TARGET_RESOLVER_CMD" with
  | Some resolver_cmd ->
      let env = Process.env_with [ ("EPI_TARGET", target) ] in
      let process_result =
        Process.run ~env ~prog:"/bin/sh" ~args:[ "-c"; resolver_cmd ] ()
      in
      if process_result.status <> 0 then
        Error
          {
            target;
            details =
              (if process_result.stderr = "" then "<no stderr>"
               else process_result.stderr);
            exit_code = Some process_result.status;
          }
      else (
        match Yojson.Basic.from_string process_result.stdout with
        | json -> Ok (descriptor_of_json json)
        | exception Yojson.Json_error msg ->
            Error { target; details = "invalid JSON output: " ^ msg; exit_code = None })
  | None -> (
      match check_target_exists target with
      | Error _ as e -> e
      | Ok () ->
          let process_result =
            Process.run ~prog:"nix"
              ~args:[ "eval"; "--json"; target ^ ".config.epi" ]
              ()
          in
          if process_result.status <> 0 then
            Error
              {
                target;
                details =
                  (if process_result.stderr = "" then "<no stderr>"
                   else process_result.stderr);
                exit_code = Some process_result.status;
              }
          else (
            match Yojson.Basic.from_string process_result.stdout with
            | json -> Ok (descriptor_of_json json)
            | exception Yojson.Json_error msg ->
                Error { target; details = "invalid JSON output: " ^ msg; exit_code = None }))

let build_target_artifact_if_missing ~target ~label =
  match split_target target with
  | None -> ()
  | Some (flake_ref, config_name) ->
      let build_target =
        match label with
        | "kernel" ->
            flake_ref ^ "#" ^ config_name ^ ".config.system.build.kernel"
        | "disk" ->
            flake_ref ^ "#" ^ config_name ^ ".config.system.build.image"
        | "initrd" ->
            flake_ref ^ "#" ^ config_name
            ^ ".config.system.build.initialRamdisk"
        | _ -> ""
      in
      if build_target <> "" then
        let _ = Process.run ~prog:"nix" ~args:[ "build"; build_target ] () in
        ()

let validate_file ~target ~label path =
  if path = "" then Error ("missing launch input: " ^ label)
  else (
    if not (Sys.file_exists path) then (
      Printf.printf "vm: building %s\n%!" label;
      ensure_store_realized path);
    if not (Sys.file_exists path) then
      build_target_artifact_if_missing ~target ~label;
    if not (Sys.file_exists path) then ensure_store_realized path;
    if not (Sys.file_exists path) then
      Error ("missing launch input: " ^ label ^ " (" ^ path ^ ")")
    else Ok ())

let is_nix_store_path path = String.starts_with ~prefix:"/nix/store/" path

let descriptor_paths descriptor =
  let base_paths = [ descriptor.kernel; descriptor.disk ] in
  match descriptor.initrd with
  | Some initrd_path -> initrd_path :: base_paths
  | None -> base_paths

let all_paths_share_parent descriptor =
  match descriptor_paths descriptor with
  | [] -> true
  | first_path :: rest ->
      let first_parent = Filename.dirname first_path in
      List.for_all
        (fun path -> String.equal (Filename.dirname path) first_parent)
        rest

let validate_descriptor_coherence descriptor =
  let kernel_is_store = is_nix_store_path descriptor.kernel in
  let disk_is_store = is_nix_store_path descriptor.disk in
  let initrd_is_store =
    match descriptor.initrd with
    | Some initrd_path -> is_nix_store_path initrd_path
    | None -> true
  in
  let any_store_path =
    kernel_is_store || disk_is_store
    || match descriptor.initrd with Some _ -> initrd_is_store | None -> false
  in
  if
    any_store_path
    && ((not kernel_is_store) || (not disk_is_store) || not initrd_is_store)
  then
    Error
      "launch inputs are not coherent: kernel/initrd/disk must all come from \
       target outputs in /nix/store; fix target outputs and rebuild instead of \
       reusing an external mutable disk image"
  else if (not any_store_path) && not (all_paths_share_parent descriptor) then
    Error
      "launch inputs are not coherent: kernel/initrd/disk must come from the \
       same target-built output set; fix target outputs and rebuild instead of \
       reusing an external mutable disk image"
  else Ok ()

let validate_descriptor ~target descriptor =
  let ( let* ) = Result.bind in
  let* () = validate_file ~target ~label:"kernel" descriptor.kernel in
  let* () = validate_file ~target ~label:"disk" descriptor.disk in
  let* () = match descriptor.initrd with
    | Some initrd_path -> validate_file ~target ~label:"initrd" initrd_path
    | None -> Ok ()
  in
  if descriptor.cpus <= 0 then Error "missing launch input: cpus must be > 0"
  else if descriptor.memory_mib <= 0 then Error "missing launch input: memory_mib must be > 0"
  else validate_descriptor_coherence descriptor

let cache_dir () =
  let dir =
    match Sys.getenv_opt "EPI_CACHE_DIR" with
    | Some dir -> dir
    | None -> (
        match Sys.getenv_opt "HOME" with
        | Some home -> Filename.concat home ".cache/epi"
        | None -> ".epi/cache")
  in
  let dir =
    if Filename.is_relative dir then Filename.concat (Sys.getcwd ()) dir
    else dir
  in
  if not (Sys.file_exists dir) then Unix.mkdir dir 0o755;
  dir

let cache_path target =
  let hash = Digest.string target |> Digest.to_hex in
  let target_cache = Filename.concat (cache_dir ()) "targets" in
  if not (Sys.file_exists target_cache) then Unix.mkdir target_cache 0o755;
  Filename.concat target_cache (hash ^ ".descriptor")

let descriptor_to_json descriptor =
  let hooks_fields =
    (match descriptor.hooks.post_launch with
     | [] -> []
     | paths -> [("post-launch", `List (List.map (fun s -> `String s) paths))])
    @
    (match descriptor.hooks.pre_stop with
     | [] -> []
     | paths -> [("pre-stop", `List (List.map (fun s -> `String s) paths))])
  in
  `Assoc ([
    ("kernel", `String descriptor.kernel);
    ("disk", `String descriptor.disk);
    ("initrd", match descriptor.initrd with Some s -> `String s | None -> `Null);
    ("cmdline", `String descriptor.cmdline);
    ("cpus", `Int descriptor.cpus);
    ("memory_mib", `Int descriptor.memory_mib);
    ("configuredUsers", `List (List.map (fun s -> `String s) descriptor.configured_users));
  ] @ (match hooks_fields with [] -> [] | _ -> [("hooks", `Assoc hooks_fields)]))

let save_descriptor_cache target descriptor =
  let path = cache_path target in
  Util.ensure_parent_dir path;
  Yojson.Basic.to_file path (descriptor_to_json descriptor)

let load_descriptor_cache target =
  let path = cache_path target in
  match Yojson.Basic.from_file path with
  | json -> Some (descriptor_of_json json)
  | exception _ -> None

let descriptor_paths_exist descriptor =
  List.for_all Sys.file_exists (descriptor_paths descriptor)

type cache_result = Cached of descriptor | Resolved of descriptor

let resolve_descriptor_cached ~rebuild target =
  let path = cache_path target in
  if rebuild && Sys.file_exists path then Sys.remove path;
  match load_descriptor_cache target with
  | Some descriptor when descriptor_paths_exist descriptor ->
      Ok (Cached descriptor)
  | _ -> (
      match resolve_descriptor target with
      | Error _ as error -> error
      | Ok descriptor ->
          save_descriptor_cache target descriptor;
          Ok (Resolved descriptor))

File: src/target.rs
use anyhow::{Context, Result, bail};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

use crate::process;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct HooksDescriptor {
    #[serde(default, alias = "post-launch")]
    pub post_launch: BTreeMap<String, String>,
    #[serde(default, alias = "pre-stop")]
    pub pre_stop: BTreeMap<String, String>,
    #[serde(default, alias = "guest-init")]
    pub guest_init: BTreeMap<String, String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Descriptor {
    pub kernel: String,
    pub disk: String,
    #[serde(default)]
    pub initrd: Option<String>,
    #[serde(default = "default_cmdline")]
    pub cmdline: String,
    #[serde(default = "default_cpus")]
    pub cpus: u32,
    #[serde(default = "default_memory_mib")]
    pub memory_mib: u32,
    #[serde(default, alias = "configuredUsers")]
    pub configured_users: Vec<String>,
    #[serde(default)]
    pub hooks: HooksDescriptor,
}

fn default_cmdline() -> String {
    "console=ttyS0 root=/dev/vda2 ro".to_string()
}

fn default_cpus() -> u32 {
    1
}

fn default_memory_mib() -> u32 {
    1024
}

impl HooksDescriptor {
    /// Sorted hook script paths for a given hook point.
    pub fn post_launch_scripts(&self) -> Vec<String> {
        self.post_launch.values().cloned().collect()
    }

    pub fn pre_stop_scripts(&self) -> Vec<String> {
        self.pre_stop.values().cloned().collect()
    }

    pub fn guest_init_scripts(&self) -> Vec<String> {
        self.guest_init.values().cloned().collect()
    }
}

/// Expand tilde in the flake-ref portion of a target string
pub fn expand_tilde(target: &str) -> String {
    if let Some((flake, config)) = target.split_once('#') {
        if let Some(rest) = flake.strip_prefix("~/")
            && let Ok(home) = std::env::var("HOME")
        {
            return format!("{home}/{rest}#{config}");
        }
        if flake == "~"
            && let Ok(home) = std::env::var("HOME")
        {
            return format!("{home}#{config}");
        }
    }
    target.to_string()
}

/// Validate target format: must contain '#'
pub fn validate(target: &str) -> Result<()> {
    let parts: Vec<&str> = target.splitn(2, '#').collect();
    if parts.len() != 2 || parts[0].is_empty() || parts[1].is_empty() {
        bail!("invalid target format: expected <flake-ref>#<config-name>, got: {target}");
    }
    Ok(())
}

/// Canonicalize target by adding nixosConfigurations. prefix if needed
pub fn canonicalize(target: &str) -> String {
    if let Some((flake, config)) = target.split_once('#')
        && !config.starts_with("nixosConfigurations.")
    {
        return format!("{flake}#nixosConfigurations.{config}");
    }
    target.to_string()
}

/// Resolve a flake target to a descriptor by evaluating nix
pub fn resolve_descriptor(target: &str) -> Result<Descriptor> {
    if let Ok(resolver) = std::env::var("EPI_TARGET_RESOLVER_CMD") {
        let out = process::run_with_env(&resolver, &[], &[("EPI_TARGET", target)])?;
        if !out.success() {
            bail!(
                "target resolver {resolver} failed (exit {}): {}",
                out.status,
                out.stderr
            );
        }
        let desc: Descriptor = serde_json::from_str(&out.stdout)
            .context("failed to parse resolver output as descriptor")?;
        return Ok(desc);
    }

    let canonical = canonicalize(target);

    // Check target exists
    let check = process::run("nix", &["eval", &canonical, "--apply", "x: true"])?;
    if !check.success() {
        bail!(
            "target resolution failed for {target} (exit {}): {}",
            check.status,
            check.stderr
        );
    }

    // Evaluate config
    let epi_attr = format!("{canonical}.config.epi");
    let eval = process::run("nix", &["eval", "--json", &epi_attr])?;
    if !eval.success() {
        bail!(
            "failed to evaluate {epi_attr} (exit {}): {}",
            eval.status,
            eval.stderr
        );
    }

    let desc: Descriptor =
        serde_json::from_str(&eval.stdout).context("failed to parse descriptor from nix eval")?;
    Ok(desc)
}

pub fn is_nix_store_path(path: &str) -> bool {
    path.starts_with("/nix/store/")
}

/// Ensure all nix store paths from the descriptor are built.
/// If any are missing, builds the flake target to produce them.
pub fn ensure_paths_exist(target: &str, desc: &Descriptor) -> Result<()> {
    let paths = descriptor_store_paths(desc);
    let any_missing = paths.iter().any(|p| !Path::new(p).exists());

    if any_missing {
        let canonical = canonicalize(target);
        let toplevel = format!("{canonical}.config.system.build.toplevel");
        let image = format!("{canonical}.config.system.build.image");
        let out = process::run("nix", &["build", &toplevel, &image, "--no-link"])?;
        if !out.success() {
            bail!("nix build failed (exit {}): {}", out.status, out.stderr);
        }
    }

    for path in &paths {
        if !Path::new(path).exists() {
            bail!("path does not exist after build: {path}");
        }
    }
    Ok(())
}

pub fn validate_descriptor(desc: &Descriptor) -> Result<()> {
    if desc.cpus == 0 {
        bail!("descriptor cpus must be > 0");
    }
    if desc.memory_mib == 0 {
        bail!("descriptor memory_mib must be > 0");
    }
    Ok(())
}

fn descriptor_store_paths(desc: &Descriptor) -> Vec<&str> {
    let mut paths = vec![desc.kernel.as_str(), desc.disk.as_str()];
    if let Some(ref initrd) = desc.initrd {
        paths.push(initrd.as_str());
    }
    for script in desc.hooks.post_launch.values() {
        if is_nix_store_path(script) {
            paths.push(script.as_str());
        }
    }
    for script in desc.hooks.pre_stop.values() {
        if is_nix_store_path(script) {
            paths.push(script.as_str());
        }
    }
    for script in desc.hooks.guest_init.values() {
        if is_nix_store_path(script) {
            paths.push(script.as_str());
        }
    }
    paths
}

// --- Caching ---

fn cache_dir() -> PathBuf {
    if let Ok(dir) = std::env::var("EPI_CACHE_DIR") {
        return PathBuf::from(dir);
    }
    if let Ok(home) = std::env::var("HOME") {
        return PathBuf::from(home).join(".cache/epi");
    }
    PathBuf::from(".epi/cache")
}

fn target_cache_path(target: &str) -> PathBuf {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};
    let mut hasher = DefaultHasher::new();
    target.hash(&mut hasher);
    let hash = hasher.finish();
    cache_dir()
        .join("targets")
        .join(format!("{hash:016x}.descriptor"))
}

pub enum CacheResult {
    Cached(Descriptor),
    Resolved(Descriptor),
}

impl CacheResult {
    pub fn descriptor(&self) -> &Descriptor {
        match self {
            CacheResult::Cached(d) | CacheResult::Resolved(d) => d,
        }
    }
}

/// Deserialize a descriptor from JSON (useful for testing and resolver output)
pub fn descriptor_from_json(json: &str) -> Result<Descriptor> {
    serde_json::from_str(json).context("failed to parse descriptor JSON")
}

pub fn resolve_descriptor_cached(target: &str, rebuild: bool) -> Result<CacheResult> {
    let cache_path = target_cache_path(target);

    if rebuild {
        let _ = fs::remove_file(&cache_path);
    }

    // Try loading from cache
    if let Ok(content) = fs::read_to_string(&cache_path)
        && let Ok(desc) = serde_json::from_str::<Descriptor>(&content)
    {
        // Verify cached paths still exist
        let paths_exist = descriptor_store_paths(&desc)
            .iter()
            .all(|p| Path::new(p).exists());
        if paths_exist {
            return Ok(CacheResult::Cached(desc));
        }
    }

    // Resolve fresh
    let desc = resolve_descriptor(target)?;

    // Cache it
    if let Some(parent) = cache_path.parent() {
        fs::create_dir_all(parent)?;
    }
    let content = serde_json::to_string_pretty(&desc)?;
    fs::write(&cache_path, content)?;

    Ok(CacheResult::Resolved(desc))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn validate_accepts_valid_target() {
        assert!(validate(".#myconfig").is_ok());
        assert!(validate("github:user/repo#config").is_ok());
    }

    #[test]
    fn validate_rejects_missing_hash() {
        assert!(validate("nohash").is_err());
    }

    #[test]
    fn validate_rejects_empty_parts() {
        assert!(validate("#config").is_err());
        assert!(validate("flake#").is_err());
        assert!(validate("#").is_err());
    }

    #[test]
    fn canonicalize_adds_prefix() {
        assert_eq!(canonicalize(".#dev"), ".#nixosConfigurations.dev");
    }

    #[test]
    fn canonicalize_preserves_existing_prefix() {
        assert_eq!(
            canonicalize(".#nixosConfigurations.dev"),
            ".#nixosConfigurations.dev"
        );
    }

    #[test]
    fn canonicalize_handles_full_flake_ref() {
        assert_eq!(
            canonicalize("github:user/repo#myvm"),
            "github:user/repo#nixosConfigurations.myvm"
        );
    }

    #[test]
    fn is_nix_store_path_works() {
        assert!(is_nix_store_path("/nix/store/abc-thing/bzImage"));
        assert!(!is_nix_store_path("/home/user/disk.img"));
    }

    #[test]
    fn validate_descriptor_rejects_zero_cpus() {
        let desc = Descriptor {
            kernel: "/k".into(),
            disk: "/d".into(),
            initrd: None,
            cmdline: "".into(),
            cpus: 0,
            memory_mib: 512,
            configured_users: vec![],
            hooks: HooksDescriptor::default(),
        };
        assert!(validate_descriptor(&desc).is_err());
    }

    #[test]
    fn validate_descriptor_rejects_zero_memory() {
        let desc = Descriptor {
            kernel: "/k".into(),
            disk: "/d".into(),
            initrd: None,
            cmdline: "".into(),
            cpus: 2,
            memory_mib: 0,
            configured_users: vec![],
            hooks: HooksDescriptor::default(),
        };
        assert!(validate_descriptor(&desc).is_err());
    }

    #[test]
    fn validate_descriptor_accepts_valid() {
        let desc = Descriptor {
            kernel: "/k".into(),
            disk: "/d".into(),
            initrd: None,
            cmdline: "".into(),
            cpus: 2,
            memory_mib: 1024,
            configured_users: vec![],
            hooks: HooksDescriptor::default(),
        };
        assert!(validate_descriptor(&desc).is_ok());
    }

    #[test]
    fn descriptor_deserialize_defaults() {
        let json = r#"{"kernel": "/k", "disk": "/d"}"#;
        let desc: Descriptor = serde_json::from_str(json).unwrap();
        assert_eq!(desc.cpus, 1);
        assert_eq!(desc.memory_mib, 1024);
        assert_eq!(desc.cmdline, "console=ttyS0 root=/dev/vda2 ro");
        assert!(desc.initrd.is_none());
        assert!(desc.configured_users.is_empty());
    }

    #[test]
    fn descriptor_deserialize_full() {
        let json = r#"{
            "kernel": "/nix/store/abc/bzImage",
            "disk": "/nix/store/abc/image.img",
            "initrd": "/nix/store/abc/initrd",
            "cmdline": "console=ttyS0",
            "cpus": 4,
            "memory_mib": 2048,
            "configured_users": ["root", "admin"],
            "hooks": {
                "post_launch": {"00-hook": "/nix/store/hook1"},
                "pre_stop": {}
            }
        }"#;
        let desc: Descriptor = serde_json::from_str(json).unwrap();
        assert_eq!(desc.cpus, 4);
        assert_eq!(desc.memory_mib, 2048);
        assert_eq!(desc.initrd.unwrap(), "/nix/store/abc/initrd");
        assert_eq!(desc.configured_users.len(), 2);
        assert_eq!(desc.hooks.post_launch.len(), 1);
    }

    #[test]
    fn descriptor_deserialize_camel_case_aliases() {
        let json = r#"{
            "kernel": "/k",
            "disk": "/d",
            "configuredUsers": ["root", "admin"],
            "hooks": {
                "post-launch": {"00-hook1": "/nix/store/hook1"},
                "pre-stop": {"00-hook2": "/nix/store/hook2"}
            }
        }"#;
        let desc: Descriptor = serde_json::from_str(json).unwrap();
        assert_eq!(desc.configured_users, vec!["root", "admin"]);
        assert_eq!(desc.hooks.post_launch_scripts(), vec!["/nix/store/hook1"]);
        assert_eq!(desc.hooks.pre_stop_scripts(), vec!["/nix/store/hook2"]);
    }

    #[test]
    fn expand_tilde_in_target() {
        let home = std::env::var("HOME").unwrap();
        assert_eq!(expand_tilde("~/repo#config"), format!("{home}/repo#config"));
        assert_eq!(expand_tilde("~#config"), format!("{home}#config"));
        assert_eq!(expand_tilde(".#config"), ".#config");
        assert_eq!(expand_tilde("/abs/path#config"), "/abs/path#config");
    }

    #[test]
    fn descriptor_roundtrip_json() {
        let desc = Descriptor {
            kernel: "/k".into(),
            disk: "/d".into(),
            initrd: Some("/i".into()),
            cmdline: "boot".into(),
            cpus: 2,
            memory_mib: 1024,
            configured_users: vec!["root".into()],
            hooks: HooksDescriptor::default(),
        };
        let json = serde_json::to_string(&desc).unwrap();
        let parsed: Descriptor = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed.kernel, desc.kernel);
        assert_eq!(parsed.cpus, desc.cpus);
    }
}
